bool goal_state
---
bool new_state